core/include/detray/propagator/rk_stepper.ipp

```diff
@@ -0,0 +1,275 @@
+
+    dk1dT = qop * vector_helpers<scalar>().cross(dk1dT, sd.b_first);
+    dk2dT = dk2dT + half_h * dk1dT;
+    dk2dT = qop * vector_helpers<scalar>().cross(dk2dT, sd.b_middle);
+    dk3dT = dk3dT + half_h * dk2dT;
+    dk3dT = qop * vector_helpers<scalar>().cross(dk3dT, sd.b_middle);
+    dk4dT = dk4dT + h * dk3dT;
+    dk4dT = qop * vector_helpers<scalar>().cross(dk4dT, sd.b_last);
+
+    // dFdT and dGdT are top-left 3x3 submatrix of equation (17)
+    auto dFdT = matrix_operator().template identity<3, 3>();
+    auto dGdT = matrix_operator().template identity<3, 3>();
+    dFdT = dFdT + h / 6. * (dk1dT + dk2dT + dk3dT);
+    dFdT = h * dFdT;
+    dGdT = dGdT + h / 6. * (dk1dT + 2. * (dk2dT + dk3dT) + dk4dT);
+
+    // Calculate dk{n}dL where L is qop
+    vector3 dk1dL = vector::cross(dir, sd.b_first);
+    vector3 dk2dL = vector::cross(dir + half_h * sd.k1, sd.b_middle) +
+                    qop * half_h * vector::cross(dk1dL, sd.b_middle);
+    vector3 dk3dL = vector::cross(dir + half_h * sd.k2, sd.b_middle) +
+                    qop * half_h * vector::cross(dk2dL, sd.b_middle);
+    vector3 dk4dL = vector::cross(dir + h * sd.k3, sd.b_last) +
+                    qop * h * vector::cross(dk3dL, sd.b_last);
+
+    // dFdL and dGdL are top-right 3x1 submatrix of equation (17)
+    vector3 dFdL = (h * h) / 6. * (dk1dL + dk2dL + dk3dL);
+    vector3 dGdL = h / 6. * (dk1dL + 2. * (dk2dL + dk3dL) + dk4dL);
+
+    // Set transport matrix (D) and update Jacobian transport
+    //( JacTransport = D * JacTransport )
+    auto D = matrix_operator().template identity<e_free_size, e_free_size>();
+    matrix_operator().set_block(D, dFdT, 0, 4);
+    matrix_operator().set_block(D, dFdL, 0, 7);
+    matrix_operator().set_block(D, dGdT, 4, 4);
+    matrix_operator().set_block(D, dGdL, 4, 7);
+
+    /// Calculate (4,4) element of equation (17)
+    /// NOTE: Let's skip this element for the moment
+    /// const auto p = getter::norm(track.mom());
+    /// matrix_operator().element(D, 3, 7) =
+    /// h * mass * mass * qop * getter::perp(vector2{1, mass / p});
+
+    this->_jac_transport = D * this->_jac_transport;
+}
+
+template <typename magnetic_field_t, typename track_t, typename constraint_t,
+          template <typename, std::size_t> class array_t>
+auto detray::rk_stepper<magnetic_field_t, track_t, constraint_t,
+                        array_t>::state::evaluate_k(const vector3& b_field,
+                                                    const int i, const scalar h,
+                                                    const vector3& k_prev)
+    -> vector3 {
+    auto& track = this->_track;
+
+    auto qop = track.qop();
+    auto dir = track.dir();
+
+    vector3 k_new;
+
+    if (i == 0) {
+        k_new = qop * vector::cross(dir, b_field);
+    } else {
+        k_new = qop * vector::cross(dir + h * k_prev, b_field);
+    }
+
+    return k_new;
+}
+
+template <typename magnetic_field_t, typename track_t, typename constraint_t,
```

Comment:
I like this, it makes the flow very easy to understand with the `advance_track(...)` and `advance_covariance(...)` calls.